File: go/modcdp/transport/PipeUpstreamTransport.go
package transport

import (
	"fmt"
	"os"
	"sync"

	"github.com/browserbase/modcdp/go/modcdp/launcher"
)

type PipeUpstreamTransport struct {
	UpstreamTransport
	PipeRead  *os.File
	PipeWrite *os.File
	writeMu   sync.Mutex
	stateMu   sync.Mutex
	closeCh   chan struct{}
}

func NewPipeUpstreamTransport(options UpstreamTransportOptions) *PipeUpstreamTransport {
	return &PipeUpstreamTransport{UpstreamTransport: NewUpstreamTransport(options), PipeRead: options.UpstreamPipeRead, PipeWrite: options.UpstreamPipeWrite}
}

func (t *PipeUpstreamTransport) Update(config map[string]any) {
	t.UpstreamTransport.Update(config)
	if config == nil {
		return
	}
	if pipeRead, _ := config["upstream_pipe_read"].(*os.File); pipeRead != nil {
		t.PipeRead = pipeRead
	}
	if pipeWrite, _ := config["upstream_pipe_write"].(*os.File); pipeWrite != nil {
		t.PipeWrite = pipeWrite
	}
}

func (t *PipeUpstreamTransport) ConfigForLauncher() LaunchOptions {
	return LaunchOptions{LauncherLocalCDPTransport: "pipe"}
}

func (t *PipeUpstreamTransport) Connect() error {
	if t.PipeRead == nil || t.PipeWrite == nil {
		return fmt.Errorf("upstream.upstream_mode=pipe requires launcher-provided pipe_read and pipe_write handles")
	}
	t.stateMu.Lock()
	if t.closeCh != nil {
		t.stateMu.Unlock()
		return nil
	}
	closeCh := make(chan struct{})
	t.closeCh = closeCh
	t.stateMu.Unlock()
	go t.readLoop(closeCh)
	return nil
}

func (t *PipeUpstreamTransport) Send(message map[string]any) error {
	t.stateMu.Lock()
	closeCh := t.closeCh
	t.stateMu.Unlock()
	if t.PipeWrite == nil || closeCh == nil {
		return fmt.Errorf("CDP pipe is not connected")
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	return launcher.WritePipeMessage(t.PipeWrite, message)
}

func (t *PipeUpstreamTransport) Close() error {
	t.stateMu.Lock()
	closeCh := t.closeCh
	t.closeCh = nil
	t.stateMu.Unlock()
	if closeCh != nil {
		close(closeCh)
	}
	if t.PipeRead != nil {
		_ = t.PipeRead.Close()
	}
	if t.PipeWrite != nil {
		_ = t.PipeWrite.Close()
	}
	return nil
}

func (t *PipeUpstreamTransport) readLoop(closeCh chan struct{}) {
	for {
		message, err := launcher.ReadPipeMessage(t.PipeRead)
		if err != nil {
			t.stateMu.Lock()
			active := t.closeCh == closeCh
			if active {
				t.closeCh = nil
			}
			t.stateMu.Unlock()
			if active {
				t.emitClose(err)
			}
			return
		}
		select {
		case <-closeCh:
			return
		default:
		}
		t.emitRecv(message)
	}
}

File: python/modcdp/transport/PipeUpstreamTransport.py
from __future__ import annotations

import json
import threading
from typing import Any

from ..transport.UpstreamTransport import UpstreamTransport, UpstreamTransportOptions


class PipeUpstreamTransport(UpstreamTransport):
    upstream_mode = "pipe"

    def __init__(self, options: UpstreamTransportOptions | None = None) -> None:
        super().__init__(options)
        options = options or {}
        self.pipe_read = options.get("upstream_pipe_read")
        self.pipe_write = options.get("upstream_pipe_write")
        self._read_thread: threading.Thread | None = None

    def update(self, config: dict[str, Any] | None = None) -> "PipeUpstreamTransport":
        config = config or {}
        self.pipe_read = config.get("upstream_pipe_read") or self.pipe_read
        self.pipe_write = config.get("upstream_pipe_write") or self.pipe_write
        return self

    def configForLauncher(self) -> dict[str, Any]:
        return {"launcher_local_cdp_transport": "pipe"}

    def connect(self) -> None:
        if self.pipe_read is None or self.pipe_write is None:
            raise RuntimeError("upstream.upstream_mode=pipe requires launcher-provided CDP pipe handles.")
        if self._read_thread is not None:
            return
        self._read_thread = threading.Thread(target=self._read_loop, daemon=True)
        self._read_thread.start()

    def send(self, message: dict[str, Any]) -> None:
        if self._read_thread is None or self.pipe_write is None:
            raise RuntimeError("CDP pipe is not connected.")
        self.pipe_write.write(json.dumps(message).encode() + b"\0")
        self.pipe_write.flush()

    def close(self) -> None:
        self._read_thread = None
        for pipe in (self.pipe_read, self.pipe_write):
            try:
                if pipe is not None:
                    pipe.close()
            except Exception:
                pass

    def _read_loop(self) -> None:
        read_thread = threading.current_thread()
        buffer = b""
        try:
            while self._read_thread is read_thread and self.pipe_read is not None:
                chunk = self.pipe_read.read(1)
                if not chunk:
                    if self._read_thread is read_thread:
                        self._handle_close(RuntimeError("CDP pipe closed"))
                    break
                buffer += chunk
                if b"\0" not in buffer:
                    continue
                raw, buffer = buffer.split(b"\0", 1)
                if raw:
                    self._parse_and_emit_recv(raw)
        except Exception as error:
            if self._read_thread is read_thread:
                self._handle_close(error if isinstance(error, Exception) else Exception(str(error)))

    def _handle_close(self, error: Exception) -> None:
        self._read_thread = None
        self._emit_close(error)
